docs/source/notes/rref.rst

```diff
@@ -0,0 +1,304 @@
+the child RRef.
+
+**G2. Parent RRef will be deleted until the child RRef is confirmed by the
+owner.**
+
+In cases 2 and 3, it is possible that the owner only partially knows the RRef
+fork graph or not even knowing it at all. For example, an RRef could be
+constructed on a user, and before the owner receives any RPC call, the
+creator user might have already shared the RRef with other users, and those
+users could further share the RRef. One invariant is that the fork graph of
+any RRef is always a tree, because forking an RRef always
+creates a new ``UserRRef`` instance on the callee (except the callee is the
+owner), and hence every RRef has a single parent.
+
+The owner's view on any ``UserRRef`` in the tree has three stages:
+
+.. code::
+
+  1) unknown -> 2) known -> 3) deleted.
+
+The owner's view of the entire tree keeps changing. The owner deletes its
+``OwnerRRef`` instance when it thinks there are no living ``UserRRef``
+instances, i.e.,
+when ``OwnerRRef`` is deleted, all ``UserRRef`` instances could be either indeed
+deleted or unknown. The dangerous case is when some forks are unknown and others
+are deleted.
+
+**G2** trivially guarantees that no parent ``UserRRef`` can be deleted before
+the owner knows all of its children ``UserRRef`` instances. However, it is
+possible that the child ``UserRRef`` may be deleted before the owner knows its
+parent ``UserRRef``.
+
+Consider the following example, where the ``OwnerRRef`` forks to A, then A forks
+to Y, and Y forks to Z.:
+
+.. code::
+
+  OwnerRRef -> A -> Y -> Z
+
+If all of Z's messages, including the delete message, are processed by the
+owner before all messages from Y, the owner will learn Z's deletion before
+knowing Y. Nevertheless, this does not cause any problem. Because, at least
+one of Y's ancestors will be alive (in this case, A) and it will
+prevent the owner from deleting the ``OwnerRRef``. More specifically, if the
+owner does not know Y, A cannot be deleted due to **G2**, and the owner knows A
+as the owner is A's parent.
+
+Things get a little trickier if the RRef is created on a user:
+
+
+.. code::
+
+  OwnerRRef
+      ^
+      |
+      A -> Y -> Z
+
+
+If Z calls to_here on the ``UserRRef``, the owner at least knows A when Z is
+deleted, because otherwise, ``to_here`` wouldn't finish. If Z does not call
+to_here, it is possible that the owner receives all messages from Z before
+any message from A and Y. In this case, as the real data of the ``OwnerRRef``
+has not been created yet, there is nothing to be deleted either. It is the same
+as Z does not exist at all. Hence, it's still OK.
+
+Implementation
+--------------
+
+**G1** is implemented by sending out a delete message in ``UserRRef``
+destructor. To provide **G2**, the parent ``UserRRef`` is put into a context
+whenever it is forked, indexed by the new ``ForkId``. The parent ``UserRRef`` is
+only removed from the context when it receives an ACK from the child, and the
```

Comment:
What's ACK? Considering linking to relevant doc for users who aren't familiar or add a short explanation.